datumaro/cli/contexts/source.py

```diff
@@ -5,186 +5,122 @@
-        try:
-            project.get_source(name)
-            raise CliException("Source '%s' already exists" % name)
-        except KeyError:
-            pass
-
-        rel_local_dir = project.local_source_dir(name)
-        local_dir = osp.join(project.config.project_dir, rel_local_dir)
-        url = args.url
-        project.env.git.create_submodule(name, local_dir,
-            url=url, branch=args.branch, no_checkout=not args.checkout)
-    elif args.source_type == 'path':
-        url = osp.abspath(args.url)
-        if not osp.exists(url):
-            raise CliException("Source path '%s' does not exist" % url)
-
-        name = args.name
-        if name is None:
-            name = osp.splitext(osp.basename(url))[0]
-
-        if project.env.git.has_submodule(name):
-            raise CliException("Git submodule '%s' already exists" % name)
-
-        try:
-            project.get_source(name)
+    name = args.name
+    if name:
+        if name in project.working_tree.sources:
             raise CliException("Source '%s' already exists" % name)
-        except KeyError:
-            pass
-
-        rel_local_dir = project.local_source_dir(name)
-        local_dir = osp.join(project.config.project_dir, rel_local_dir)
-
-        if args.copy:
-            log.info("Copying from '%s' to '%s'" % (url, local_dir))
-            if osp.isdir(url):
-                # copytree requires destination dir not to exist
-                shutil.copytree(url, local_dir)
-                url = rel_local_dir
-            elif osp.isfile(url):
-                os.makedirs(local_dir)
-                shutil.copy2(url, local_dir)
-                url = osp.join(rel_local_dir, osp.basename(url))
-            else:
-                raise Exception("Expected file or directory")
+    else:
+        name = generate_next_name(
+            list(project.working_tree.sources) + os.listdir(),
+            'source', sep='-', default='1')
+
+    fmt = args.format
+    if fmt in project.working_tree.env.importers:
+        arg_parser = project.working_tree.env.importers[fmt]
+    elif fmt in project.working_tree.env.extractors:
+        arg_parser = project.working_tree.env.extractors[fmt]
+    else:
+        raise CliException("Unknown format '%s'. A format can be added"
+            "by providing an Extractor and Importer plugins" % fmt)
+
+    extra_args = {}
+    if args.extra_args:
+        if hasattr(arg_parser, 'parse_cmdline'):
+            extra_args = arg_parser.parse_cmdline(args.extra_args)
         else:
-            os.makedirs(local_dir)
+            raise CliException("Format '%s' does not accept "
+                "extra parameters" % fmt)
 
-    project.add_source(name, { 'url': url, 'format': args.format })
+    project.import_source(name, url=args.url, format=args.format,
+        options=extra_args, no_cache=args.no_cache, rpath=args.path)
+    on_error_do(project.remove_source, name, force=True, keep_data=False,
+        ignore_errors=True)
```

Comment:
It's difficult to tell which keyword arguments are meant for `on_error_do` and which for `project.remove_source`. I think that instead of having `on_error_do` accept arbitrary kwargs, it should have a `fwd_kwargs` parameter (like the one in `Rollback.add`).

Reply:
Done